### PR TITLE
feat(pgvector): cluster_vectors — in-process k-means over an embedding column (A5)

### PR DESCRIPTION
**Pull request by devopam**
## Why

Workstream A, **item A5** from `docs/parallel-roadmap.md` (shortlist **9.3**). Third resident of `mcpg.vector_ops`. Lays the centroid + assignment groundwork that **A6** (`detect_vector_outliers`, sequenced after this) will reuse — per the roadmap's intra-stream dependency note.

## What

**`cluster_vectors(schema, table, embedding_column, k, ...)`** — in-process k-means over an embedding column. Read-only.

- Samples up to `sample_size` (default **5000**) non-NULL rows; silently skips rows that don't parse to the column's declared `vector(N)` dimension.
- **k-means++ seeding** (deterministic via `seed`) + **Lloyd update loop** with a centroid-drift convergence check (tolerance `1e-4` of the mean centroid scale).
- **Empty-cluster re-seeding** on the worst-fit row prevents the classic degenerate output.
- `metric="l2"` (default, squared Euclidean) or `"cosine"` (vectors normalised up front, centroids re-normalised every iteration so the standard Lloyd update still converges).
- Returns `ClusterCentroid[]` (cluster, centroid, size) + `ClusterAssignment[]` (id or sample index, cluster, distance), plus `iterations`, `converged`, `inertia`.

### Module-level helpers (composable with A6)
- `_squared_distance`, `_cosine_distance`, `_normalize_in_place` — pure numerical helpers, all individually tested.
- `_kmeans_plus_plus_init` + `_kmeans` — algorithm guts kept separate from the I/O-bound entry point.
- `_vector_column_dimension` — sibling of the helper in `data_movement.py`, duplicated for module independence. Same shape as the helper introduced in #50 (`cross_table_similarity`); whichever lands first owns it on `main` and the rebase collapses the other.

## Test plan
- [x] **+16 tests**: `_squared_distance` / `_cosine_distance` / `_normalize_in_place` boundary cases; two well-separated clusters land cohesively (deterministic *structural* assertion, not seed-dependent); `id_column` passthrough; cosine metric yields unit-norm centroids; **same seed → identical centroids on two runs**; `sample_size` bound as `LIMIT`; wrong-dim rows silently skipped; not-enough-rows error; column-not-pgvector; arg validation (`metric` / `k=0` / `k=1` / `sample_size=0` / `max_iterations=0`); parametric identifier safety; tool listing + dispatch via MCP client.
- [x] `vector_ops.py` at **97% coverage** (the empty-cluster reseed branch + a couple of rare paths uncovered — acceptable for a heuristic algorithm without a real PG to drive it).
- [x] **1116 unit tests pass**; ruff / format / mypy(`src/mcpg`) / bandit clean.

## Docs (folded in, per parallel-roadmap §2)
- `tour.md`: tool line added under "Tune pgvector".
- `feature-shortlist.md`: 9.3 → ✅.
- `CHANGELOG.md`: `[Unreleased]` bullet.
- **Tool count NOT bumped** — parallel batch with A3 / A4 / A6; periodic sync reconciles.

## Next
**A6** (`detect_vector_outliers`) waits for this to merge so it can reuse the centroid helpers + dataclasses without a duplicate-definition collision.

---
_Generated by [Claude Code](https://claude.ai/code/session_0122yLZLJ8t4W43sdN6BmTZc)_

### CODE REVIEW

File: src/mcpg/vector_ops.py [hunk 1]
```diff
@@ -250,3 +279,342 @@ async def analyze_distance_metric(
...
+    dot = 0.0
+    norm_a = 0.0
+    norm_b = 0.0
+    for x, y in zip(a, b, strict=True):
+        dot += x * y
+        norm_a += x * x
+        norm_b += y * y
+    if norm_a == 0.0 or norm_b == 0.0:
+        return 1.0
+    return 1.0 - dot / math.sqrt(norm_a * norm_b)
```
[the remark this answers is not in the document]
Comment (devopam) [reply src/mcpg/vector_ops.py:538]: Good catch on the redundant norm work in the hot loop — fixed in 7668d0f, but kept the public `_cosine_distance` correct for the general (non-unit) case. The k-means hot loop now uses an inlined `_cosine_distance_unit` (just `1 - dot(a, b)`) since vectors and centroids are guaranteed unit-norm there (caller pre-normalises inputs; centroids are re-normalised every iteration). The public helper stays as a general-purpose cosine distance — tests import it and call it with non-unit vectors deliberately, so changing its contract would silently break them.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/mcpg/vector_ops.py [hunk 2]
```diff
@@ -250,3 +279,342 @@ async def analyze_distance_metric(
...
+    while len(centroids) < k:
+        # Squared distance from each point to its nearest existing centroid.
+        weights = []
+        total = 0.0
+        for v in vectors:
+            best = min(distance(v, c) for c in centroids)
+            # Square so larger gaps dominate the probability mass.
+            w = best * best
+            weights.append(w)
+            total += w
```
[the remark this answers is not in the document]
Comment (devopam) [reply src/mcpg/vector_ops.py]: Real algorithmic bug — fixed in 7668d0f. You're right that `_squared_distance` already returns D², so my `w = best * best` was actually sampling proportional to D⁴, heavily over-weighting outliers. Resolution: just use `w = best` (D² for the L2 path, `1 - cos(θ)` for the cosine path — both non-negative weights with sensible scale). The conditional `best if distance == _squared_distance else best * best` works too but I went with the simpler unconditional form since cosine k-means literature also typically uses the raw cosine distance (not its square) for seeding weights.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/mcpg/vector_ops.py [hunk 3]
```diff
@@ -250,3 +279,342 @@ async def analyze_distance_metric(
...
+            for d in range(dim):
+                row[d] += v[d]
+        for j in range(k):
+            if counts[j] == 0:
+                # Re-seed an empty cluster on the point currently farthest
+                # from its assigned centroid — a standard safety valve.
+                worst = max(range(len(vectors)), key=lambda i: distances[i])
+                new_centroids[j] = list(vectors[worst])
+                counts[j] = 1
+                continue
```
[the remark this answers is not in the document]
Comment (devopam) [reply src/mcpg/vector_ops.py:662]: Subtle bug, well caught — fixed in 7668d0f exactly as suggested (`distances[worst] = 0.0` after the pick). Added `test_cluster_vectors_avoids_duplicate_centroids_when_many_clusters_collapse` as a regression: runs k=4 over 8 well-separated 2D points and asserts every final centroid is distinct. With the pre-fix logic this would yield duplicate centroids on any iteration where multiple clusters collapsed (which the seeding heuristic + small datasets can produce).

---
_Generated by [Claude Code](https://claude.ai/code)_
